lib/web_ui/lib/src/engine/semantics/dialog.dart

```diff
@@ -16,6 +16,39 @@ class Dialog extends PrimaryRoleManager {
     // names its own route an `aria-label` is used instead of `aria-describedby`.
     addFocusManagement();
     addLiveRegion();
+
+    // When a route/dialog shows up it is expected that the screen reader will
+    // focus on something inside it. There could be two possibilities:
+    //
+    // 1. The framework explicitly marked a node inside the dialog as focused
+    //    via the `isFocusable` and `isFocused` flags. In this case, the node
+    //    will request focus directly and there's nothing to do on top of that.
+    // 2. No node inside the route takes focus explicitly. In this case, the
+    //    expectation is to look through all nodes in traversal order and focus
+    //    on the first one.
+    semanticsObject.owner.addOneTimePostUpdateCallback(() {
+      if (semanticsObject.owner.hasNodeRequestingFocus) {
+        // Case 1: a node requested explicit focus. Nothing extra to do.
+        return;
+      }
+
+      // Case 2: nothing requested explicit focus. Focus on the first descendant.
+      _setDefaultFocus();
```

Comment:
We may need to add a 3rd case:

3. Nothing was focused before the semantics update, and nothing requested focus during this update, so leave everything as is.

This 3rd case will be common when Flutter is embedded inside another app, where the screen reader might be focusing somewhere else outside of the Flutter app. We don't want to steal that focus, do we?

Reply:
If something in the embedded flutter view triggered a route change inside, we should bring the focus here I think.

Reply:
In other platform we send screen change accessibilityEvent and let OS decide whether to move the accessibility focus. so we are in an uncharted territory here.

Reply:
When we have focusability at the `FlutterView` level, one possible middle ground could be:

```
/// Case 3: the route popped up inside a FlutterView that's not currently
            focused. Do not auto-focus inside the route because that could
            "steal" focus from an element outside the current FlutterView.
```

In the meantime, I agree with @chunhtai that if a route change happened, it is likely because the user took some action to push that route, and the most common outcome should be that a widget inside the new route takes focus. We can revisit when we have `FlutterView` focusability (currently worked on by @tugorez).